### ISSUE TITLE
AssemblyLoadContext: Provide isolation for dependencies of an assembly loaded in ALC

### ISSUE THREAD

**Issue by gokarnm** [dotnet/coreclr#6470]
Opening a new issue as I'm not able to reopen #6269. Let me know if I should create new issues or request reopening closed issue in future.

**Context** : I'm creating a custom AssemblyLoadContext in my application to load a plugin from a directory. The plugin can have additional dependencies on CoreFX or third party libraries. These dependencies can be same as my application's dependencies but the plugin may use a different version. 

After the fix (https://github.com/dotnet/coreclr/pull/6335), I'm able to successfully load different copies of a third party dependency. I'm not able to load a different copy of a CoreFX library, or verify isolation.

Here are the scenarios I'm trying to implement using ALC :
**Scenario 1**: The plugin is developed against a patch release of .NET Core (1.0.1) whereas the application that loads the plugin was developed against .NET Core 1.0.
**Scenario 2**: The plugin is developed against a minor ([Fast Track Support](https://www.microsoft.com/net/core/support)) release of .NET Core (1.1) whereas the application that loads the plugin was developed against .NET Core 1.0.

In my test, I'm replicating Scenario 1 by copying the dependency graph of Microsoft.CSharp.dll and System.Reflection.dll from install location (C:\Program Files\dotnet\shared\Microsoft.NETCore.App\1.0.0) to the plugin bin directory. I'm assuming a patch release would have same AssemblyVersion and higher AssemblyInformationalVersion. When the application loads the plugin into a custom ALC, I see Load method of the custom ACL being invoked to resolve dependencies of the plugin, at this point the same dependencies are already loaded in the default context. If I find the dependency in the plugin bin, I load it using AssemblyLoadContext.LoadFromAssemblyPath. If I don't find the assembly in the plugin bin directory, I return null, in which case it seems like it falls back to TPA binder.

 I'm verifying isolation by
1. Checking Visual Studio->Debug->Windows->Modules to verify if multiple copies of the dependency are loaded.
2. With the third party assembly, setting a static field and verifying if the statics are independent.
3. With CoreFx dependencies I print typeof(SomeTypeFromTheLibrary).GetHashCode() in the application and the plugin to check if the Type are independent, e.g. _Console.WriteLine($"MethodInfo hash:{typeof(MethodInfo).GetHashCode()}")_. Please comment if this is a reliable/correct test.

Am seeing inconsistent behavior with different dependencies as follows:
- Third Party dependency: The Modules window shows the DLL loaded from multiple locations; nuget cache and plugin bin. The statics are independent and Type hash codes are different.
- Microsoft.CSharp : If it's not a dependency of the application, it's loaded from the plugin bin. But if it's already loaded by the application, a second copy is not loaded from plugin bin even after AssemblyLoadContext.LoadFromAssemblyPath is called in custom ALC's Load method. Type hash codes are different in default and custom ALC.
- System.Reflection: The application depends on System.Reflection so it's always present in default ALC. A second copy is not loaded from plugin bin even after AssemblyLoadContext.LoadFromAssemblyPath is called in custom ALC's Load method. Type hash codes are same in default and custom ALC.

**Questions**:
1. Can scenarios 1 & 2 be implemented using AssemblyLoadContext?
2. I'm aware of Trusted Platform Assemblies but cannot find any documentation for it. What is a TPA? Are there special loading/resolution rules for assemblies in TPA list?
3. Using ildasm on System.Reflection shows that some types are forwarded to mscorlib. E.g.
   _.class extern forwarder System.Reflection.MethodInfo
   {
   .assembly extern mscorlib
   }_.
   How does this affect versioning? Does this mean that some Core FX libraries like System.Reflection are versioned in lockstep with mscorlib? Does this mean that to satisfy scenario 1, the plugin needs to provide it's own copy of mscorlib?
4. Will the recent fix (https://github.com/dotnet/coreclr/pull/6335) be included in the [1.0.1 release ](https://blogs.msdn.microsoft.com/dotnet/2016/07/15/net-core-roadmap/)? Will you be cherry picking commits/fixes for 1.0.1 release or all commits till a particular point will be included?

**Comment by jkotas** [comment]
cc @gkhanna79

**Comment by gkhanna79** [comment]
> I'm assuming a patch release would have same AssemblyVersion and higher AssemblyInformationalVersion

AssemblyInformationalVersion has no meaning with respect to assembly loads.

> If I don't find the assembly in the plugin bin directory, I return null, in which case it seems like it falls back to TPA binder.

This is the intended, by-design behavior. 

> But if it's already loaded by the application, a second copy is not loaded from plugin bin even after AssemblyLoadContext.LoadFromAssemblyPath is called in custom ALC's Load method

Do you have a repro for this? After loading Microsoft.CSharp in the Load override, do you not get back a valid assembly or is an exception thrown? If you are getting a valid assembly, invoke AssemblyLoadContext.GetLoadContext against it - it should be the same as your custom ALC.

> A second copy is not loaded from plugin bin even after AssemblyLoadContext.LoadFromAssemblyPath is called in custom ALC's Load method.

Please share a repro for this, including doing the experiment as above. This is not expected after the fix #6335 unless Reflection is completely type-forwarded to System.Private.CoreLib (see more context below).

> Can scenarios 1 & 2 be implemented using AssemblyLoadContext?

Yes - AssemblyLoadContext is designed for it. 

> I'm aware of Trusted Platform Assemblies but cannot find any documentation for it.

TPA used to be a special set of assemblies that comprised the platform assemblies, when it was originally designed. As of today, it is simply the set of assemblies known to constitute the application, per its app.deps.json.

> Does this mean that some Core FX libraries like System.Reflection are versioned in lockstep with mscorlib?

For the case where the entire implementation is present in System.Private.CoreLib (the Core Library - mscorlib is a facade that forwards type implementations to it), then yes, they are locked together. And yes, there can only be one Core library within the process (which is next to CoreCLR.dll).

#6335 is currently planned for our fall release.

**Comment by gokarnm** [comment]
Thanks @gkhanna79 for the quick response and for answering all the questions!

>Do you have a repro for this? After loading Microsoft.CSharp in the Load override, do you not get back a valid assembly or is an exception thrown? If you are getting a valid assembly, invoke AssemblyLoadContext.GetLoadContext against it - it should be the same as your custom ALC.

I modified the test to call AssemblyLoadContext.GetLoadContext. I do get a valid assembly after calling AssemblyLoadContext.LoadFromAssemblyPath and calling AssemblyLoadContext.GetLoadContext against it returns the custom ALC. ALC is working as intended, it's the type forwarding and not seeing copies of modules being loaded in _Debug->Modules_ window that tripped me.

>This is not expected after the fix #6335 unless Reflection is completely type-forwarded to System.Private.CoreLib (see more context below).

All types in System.Reflection are forwarded. I tried the same test with types from System.IO.dll, System.IO.Stream which is forwarded ends up with same Type hash in default and custom context. System.IO.BufferedStream which is not forwarded has different Type hashes in default and custom context.

I have a few follow up questions:

>For the case where the entire implementation is present in System.Private.CoreLib (the Core Library - mscorlib is a facade that forwards type implementations to it), then yes, they are locked together. And yes, there can only be one Core library within the process (which is next to CoreCLR.dll).

1. For types that are forwarded, do they have independent statics in different load contexts? I assuming they don't as they resolve to the same Type.

2. Given that System.Reflection and quite a number of other types across CoreFX are currently forwarded, I'm assuming I cannot make scenario 1 work for cases where a plugin is developed against a patch release of .NET Core (1.0.1) and has a dependency on a type that is forwarded. When the plugin is loaded by the application the dependency will finally resolve to the implementation in System.Private.CoreLib that was loaded by the application. Is that assumption correct?

3. Why are some types from CoreFX implemented in CoreCLR? Will these type forwards change in future? Will the implementation be moved to CoreFx making libraries (like System.Reflection, System.IO) truly modular so they are not locked with CoreCLR?

4. I don't see copies of Microsoft.CSharp and System.Reflection DLLs being loaded from plugin bin in the Debug->Modules window. These DLLs in plugin bin are copies from dotnet install location. Is the reason why I don't see copies loaded is because the runtime checks assembly content hash and doesn't load it if it's already loaded?

5. Is .NET Core 1.0.1 a LTS patch release and the 1.1 a FTS release?

**Comment by gkhanna79** [comment]
@gokarnm Answers in order asked:

1) Since typed-forwarded types are always coming from the same context, statics for them come from the same context as well. Basically, statics are associated with the Loadcontext the type comes from.

2) Correct.

3) Part of the answer is legacy. However, we have been trying to move things out to the degree we can. @weshaggard can comment on specific libraries/types.

4) Are you asking if the exact same binary, but from a different location, should be loaded in your custom context or not?

5) LTS patch release.

**Comment by gokarnm** [comment]
Thanks @gkhanna79 !

>4) Are you asking if the exact same binary, but from a different location, should be loaded in your custom context or not?

Yes, what should happen if the default context has loaded a DLL and the plugin bin also contains the exact same binary? The LoadFromAssemblyPath method call in my custom ALC returns the DLL from plugin bin (the returned assembly.Locations points to DLL path in plugin bin), but Debug->Modules window does not reflect this. If it actually is loaded only once, how does the runtime detect that it's the exact same binary? I'm trying to understand the behavior in this case.

**Comment by gkhanna79** [comment]
> but Debug->Modules window does not reflect this. 

I am not sure why the Modules window does not reflect this and wonder if the OS loader is doing some mapping magic. I presume that if you fetch the load context for the loaded assembly, it represents the custom ALC, right?

**Comment by gkhanna79** [comment]
@gokarnm Any update here?

**Comment by gokarnm** [comment]
>I presume that if you fetch the load context for the loaded assembly, it represents the custom ALC, right?

Yes, the [fix](https://github.com/dotnet/coreclr/pull/6335) works correctly. Closing this issue.
